src/lxd.py

```diff
@@ -0,0 +1,321 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""LXD client interface.
+
+The Lxd class is intend to be layer of abstraction to isolate the underlying implementation of LXD.
+"""
+from __future__ import annotations
+
+import io
+import tempfile
+from subprocess import CalledProcessError  # nosec B404
+from typing import IO, Optional, Tuple, Union
+
+import pylxd.models
+
+from errors import LxdError
+from lxd_type import LxdInstanceConfig, ResourceProfileConfig, ResourceProfileDevices
+from utilities import execute_command, secure_run_subprocess
+
+
+class LxdInstanceFiles:
+    """File manager of a LXD instance.
+
+    Attrs:
+        instance (LxdInstance): LXD instance where the files are located in.
+    """
+
+    def __init__(self, instance: LxdInstance):
+        """Construct the file manager.
+
+        Args:
+            instance: LXD instance where the files are located in.
+        """
+        self.instance = instance
+
+    def mk_dir(self, dir_name: str) -> None:
+        """Create a directory in the LXD instance.
+
+        Args:
+            dir: Name of the directory to create.
+        """
+        self.instance.execute(["/usr/bin/mkdir", "-p", dir_name])
+
+    def push(self, source: str, destination: str, mode: Optional[str] = None) -> None:
+        """Push a file to the LXD instance.
+
+        Args:
+            source: Path of the file to push to the LXD instance.
+            destination: Path in the LXD instance to load the file.
+            mode: File permission setting.
+
+        Raises:
+            LxdException: Unable to load the file into the LXD instance.
+        """
+        lxc_cmd = [
+            "/snap/bin/lxc",
+            "file",
+            "push",
+            source,
+            f"{self.instance.name}/{destination.lstrip('/')}",
+        ]
+
+        if mode:
+            lxc_cmd += ["--mode", str(mode)]
+
+        try:
+            execute_command(lxc_cmd)
+        except CalledProcessError as err:
+            raise LxdError(f"Unable to push file into LXD instance {self.instance.name}") from err
+
+    def put(self, filepath: str, content: Union[str, bytes], mode: Optional[str] = None) -> None:
```

Comment:
What is the difference between `put` and `push`? I think one pushes an existing fle and the other just based on a string, are both needed?

Reply:
One is for pushing file and one is for pushing string.

The runner application is a tar file downloaded. So it is pushed by the file, rather than reading the bytes of the file and then convert to a temp file to be pushed.
Other configuration files are put into the instance with by the string, as they are a string generated from a template.